new/property.md

```diff
@@ -0,0 +1,206 @@
+
+```ruby
+property :x, must_match: [String, :a, :b, nil]
+```
+
+### Other options
+
+#### coerce
+
+`coerce` is a proc run when the user sets a non-lazy value, or reads a lazy or default value. It allows normalization of input, which makes it simple to create expressive interfaces while preserving a simple programming model that knows what to expect:
+
+```ruby
+class File < Chef::Resource
+  attribute :mode, coerce: proc { |v| mode.is_a?(String) ? mode.to_s(8) : mode }
+end
+```
+
+`coerce` procs are run in the context of the instance, so that they have access to other attributes and methods.
+
+#### default
+
+This works similarly to `attribute`, except that `lazy` values are automatically run in the context of the *instance*:
+
+```ruby
+class MysqlInstall < Chef::Resource
+  property :root_path, String, default: '/'
+  property :config_path, String, default: lazy { File.join(root_path, 'config') }
+end
+```
+
+#### name_attribute
+
+Same as before. Causes the attribute to be explicitly set to the name passed to the constructor. To wit:
+
+```ruby
+class MyResource < Chef::Resource
+  property :path, name_attribute: true
+end
+
+my_resource 'foo' do
+  name 'bar'
+  puts path #=> foo
+end
+```
+
+#### patchy
+
+Properties declare whether they are patchy (meaning resource actions will not change the on-disk value) or not patchy by specifying `patchy: true|false`. The primary effect of this is to prevent the property from being cloned during Chef's clone process (which happens when you declare a resource twice with different properties):
+
+
+```ruby
+file '/mystuff/x.txt' do
+  mode 0666
+end
+
+<long series of actions ...>
+
+execute 'chmod /mystuff 0777'
+
+<long series of actions ...>
+
+# If mode is declared `patchy: false`, we will change mode back to 0666 here.
+# If mode is declared `patchy: true`, we leave mode at 0777.
+file '/mystuff/x.txt' do
+  content 'Hello World'
+end
+```
+
+The reason being, a patchy property is one that *leaves the value alone* unless the user actually says they want to change it.  Cloning values from other resources violates that.
+
+### Property type
+
+Properties with a single type are common enough that we support a primary "type" for a property, specifiable after its name.
+
+```ruby
+class MyResource < Chef::Resource
+  property :content, String
+end
+```
+
+#### Reusing types
+
+The type of a property is represented by `Chef::Resource::PropertyType`, and contains accessors for all the properties of a type (`must_be`, `name_attribute`, `kind_of`, etc.).
+
+When you declare `property :name, <type>, <options>`, one of two things happens:
+
+- If the type is a PropertyType instance, it is dup'd and any <option>s are set on the new type.
+- If the type is a Class, a new PropertyType instance is created with `kind_of` set to [Class] and any options are set on the new type.
+- If type is not passed, a new PropertyType instance is created, and any options are set on the new type.
```

Comment:
Yeah, I was thinking maybe a `properties` on the class.  As for whether the property was set, I'd like an `is_set?(:attr_name)`, which I'm quite sure will make it into one of these RFCs (though I don't think this RFC made the need for it any *worse* than it is now).